### PR TITLE
Put the GAP version into repository

### PR DESCRIPTION
**Pull request by fingolfin**
So far, the GAP version and release date are injected into the GAP sources by
a script during the release process. This is in stark contrast to how this is
handled by (almost) all GAP packages using git.

This design makes it difficult to test certain effects of the GAP release
process as part of our CI. It also means that source snapshots of the GAP
repository have little information about the underlying version.

This patch changes things by putting the GAP version and release date into
configure.ac, from where it is distributed into all files that need them. Thus
to change either, one only needs to change a single file.

In addition, we get rid of NeedKernelVersion, as we never release GAP library
and kernel separately these days.

### CODE REVIEW

**Comment by alex-konovalov** [comment]
@fingolfin quick question - does that mean each stable-4.X branch will have different version number set there too?

**Comment by fingolfin** [comment]
@alex-konovalov yes exactly. That's explicitly something I wanted to have.

**Comment by alex-konovalov** [comment]
@fingolfin another QQ: am I right that adopting this PR will not require major changes in wrapping releases, since the only changes that will be needed is to drop this step: https://github.com/gap-system/gap-distribution/blob/master/DistributionUpdate/updateversioninfo - it seems like all things that are adjusted there (version, date, year, etc) are now handled by this PR?

**Comment by fingolfin** [comment]
@alex-konovalov that is the intention, yes

File: .gitignore [hunk 1]
```diff
@@ -52,6 +54,7 @@ main.out
 
 # Documentation
 /doc/make_doc
+/doc/version.data
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply .gitignore]: Actually, I refactored this PR a lot and now `cnf/GAP-VERSION-GEN` is back. So only `doc/versiondata` is being renamed. 

And I *could* avoid that, if we are willing to have `/doc/versiondata` in the repository and that after updating the version in `configure.ac`, you are supposed to remember to re-run the build system, then commit the changes it produces in `doc/versiondata` -- I am already expecting you to do that with `CITATION`, so I guess it'd be OK?

**Comment by fingolfin** [comment]
An alternative approach would be to add a helper script, say `etc/update_version` which takes the desired version, release year and  release date/day as argument, and then uses `sed` to update all relevant files from some template. The advantage of this is that updating the version then just is two steps (assuming you start from a clean clone of the GAP repository):
 1. run `etc/update_version 4.12.1  2020-02-24`  (assuming we give an ISO date and let it extract the year)
 2. commit all modified files.

While with the current approach, it requires three, and it is easy to forget the middle step (though Travis CI would immediately catch it)
  1. update version etc. in `configure.ac` with an editor
  2. run `make` (resp. `./autogen.sh && ./configure && make`)
  3. commit all modified files.

**Comment by fingolfin** [comment]
I realized that one already had to use the three step approach with my previous iteration of this PR, because of the CITATION file, so for now I went "all in" on that to reduce the size of the patch.

Moreover, as suggested by @ChrisJefferson I adjusted the PR to show the build date&time when starting a non-release build, instead of the string "today".

**Comment by fingolfin** [comment]
@alex-konovalov @ChrisJefferson I suggest we backport this to 4.11 and then immediately update (or rather: remove) https://github.com/gap-system/gap-distribution/blob/master/DistributionUpdate/updateversioninfo

**Comment by ChrisJefferson** [comment]
I'm happy to merge this, but will leave a while for any final comments (in particular from @alex-konovalov -- if you are happy with it, feel free to merge)

**Comment by alex-konovalov** [comment]
Thanks. Anyone willing to submit a PR to update https://github.com/gap-system/gap-distribution/blob/master/DistributionUpdate/updateversioninfo ?

**Comment by fingolfin** [comment]
@alex-konovalov That script should simply be removed. The actual work that is required is to adapt the release checklist to say that one ought to commit the new version to the GAP git repository. I could look into this, but overall, I think that'd be a waste of time -- the plan is to discard all that during GAP Days anyway, and redo it from scratch. And for 4.11, we'd still use the current system (this PR won't be in stable-4.11, after all).

**Comment by fingolfin** [comment]
And yes, that runs contrary to my earlier suggestion to "backport" this. I now think this was a misleaded idea. It will only delay GAP 4.11 further with no gain. Let's use the well-established release system one last time for that.

**Comment by fingolfin** [comment]
Note that we might also want to make GAP 4.11.1, 4.11.2, ... and then those all would likely have to use the old release system, *unless* we port all required changes to the stable-4.11 branch. Which we could ostensibly do, but I'd suggest to first complete the new release system, and test it thoroughly. And then decide whether to wait for GAP 4.12. Which normally should be released around May, though I guess we should postpone that; how much depends on when exactly GAP 4.11 will be released.

This all assumes that a GAP 4.11 release happens in February. If that doesn't work out again, I expect that we'll create the release during the GAP Days in March, and then of course this PR would be backported to stable-4.11 right away -- but in parallel to the creation of the new scripts.

**Comment by alex-konovalov** [comment]
I agree with @fingolfin 
 
> And yes, that runs contrary to my earlier suggestion to "backport" this. I now think this was a misleaded idea. It will only delay GAP 4.11 further with no gain.

**Comment by fingolfin** [comment]
Just to clarify: Now that 4.11.0 is out, we could still backport this for use with 4.11.1 . Or not, depending on what seems most convenient

**Comment by fingolfin** [comment]
I'd like to backport this to stable-4.11, but only once the new release system is ready; also, if we backport this, we must make sure to also backport PR #3927 which fixes an issue in this PR.

**Comment by alex-konovalov** [comment]
@fingolfin do not backport this to stable-4.11 yet, please. 

I am working now on changes in https://github.com/gap-system/gap-distribution/tree/master/DistributionUpdate and need to first make them and test that all works for GAP 4.12.0 release candidate. While that, GAP 4.11.1 RC will be wrapped from https://github.com/gap-system/gap-distribution/tree/stable-4.11/DistributionUpdate branch. Well all done, we need to simultaneously backport this PR to master, and switch wrapping 4.11 series in a matching way.

**Comment by alex-konovalov** [comment]
@fingolfin could you backport this and #3927 to stable-4.11 now, please?

**Pull request merged**